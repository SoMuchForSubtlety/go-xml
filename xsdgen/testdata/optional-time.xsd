<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:ns="http://example.org"
  targetNamespace="http://example.org">

  <xs:element name="Type1">
    <xs:complexType>
      <xs:sequence>
        <xs:element maxOccurs="1" minOccurs="0" name="optionalTime" nillable="true"
          type="xs:dateTime"> </xs:element>
        <xs:element maxOccurs="1" minOccurs="1" name="requiredTime" nillable="false"
          type="xs:dateTime"> </xs:element>
      </xs:sequence>
    </xs:complexType>
  </xs:element>

</xs:schema>
