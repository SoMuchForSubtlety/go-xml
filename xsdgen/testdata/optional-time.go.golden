// Code generated by xsdgen.test. DO NOT EDIT.

package ws

import (
	"bytes"
	"encoding/xml"
	"time"
)

type Type1 struct {
	OptionalTime time.Time `xml:"http://example.org optionalTime,omitempty"`
	RequiredTime time.Time `xml:"http://example.org requiredTime"`
}

func (t *Type1) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
	type T Type1
	var overlay struct {
		*T
		OptionalTime *xsdDateTime `xml:"http://example.org optionalTime,omitempty"`
		RequiredTime *xsdDateTime `xml:"http://example.org requiredTime"`
	}
	overlay.T = (*T)(t)
	overlay.OptionalTime = (*xsdDateTime)(&overlay.T.OptionalTime)
	overlay.RequiredTime = (*xsdDateTime)(&overlay.T.RequiredTime)
	return d.DecodeElement(&overlay, &start)
}
func (t Type1) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
	type T Type1
	var layout struct {
		*T
		OptionalTime *xsdDateTime `xml:"http://example.org optionalTime,omitempty"`
		RequiredTime *xsdDateTime `xml:"http://example.org requiredTime"`
	}
	layout.T = (*T)(&t)
	layout.OptionalTime = (*xsdDateTime)(&layout.T.OptionalTime)
	layout.RequiredTime = (*xsdDateTime)(&layout.T.RequiredTime)
	return e.EncodeElement(layout, start)
}

type xsdDateTime time.Time

func (t *xsdDateTime) UnmarshalText(text []byte) error {
	return _unmarshalTime(text, (*time.Time)(t), "2006-01-02T15:04:05.999999999")
}
func (t xsdDateTime) MarshalText() ([]byte, error) {
	return _marshalTime((time.Time)(t), "2006-01-02T15:04:05.999999999")
}
func (t xsdDateTime) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
	if (time.Time)(t).IsZero() {
		return nil
	}
	m, err := t.MarshalText()
	if err != nil {
		return err
	}
	return e.EncodeElement(m, start)
}
func (t xsdDateTime) MarshalXMLAttr(name xml.Name) (xml.Attr, error) {
	if (time.Time)(t).IsZero() {
		return xml.Attr{}, nil
	}
	m, err := t.MarshalText()
	return xml.Attr{Name: name, Value: string(m)}, err
}
func _unmarshalTime(text []byte, t *time.Time, format string) (err error) {
	s := string(bytes.TrimSpace(text))
	*t, err = time.Parse(format, s)
	if _, ok := err.(*time.ParseError); ok {
		*t, err = time.Parse(format+"Z07:00", s)
	}
	return err
}
func _marshalTime(t time.Time, format string) ([]byte, error) {
	return []byte(t.Format(format + "Z07:00")), nil
}
